docs/doxygen/man/man3/GetBroadcastMsg.3
.TH "GetBroadcastMsg" 3 "Thu Aug 30 2018" "esc" \" -*- nroff -*-
.ad l
.nh
.SH NAME
GetBroadcastMsg \- Get broadcast command class\&.  

.SH SYNOPSIS
.br
.PP
.PP
\fC#include <getbroadcastmsg\&.h>\fP
.PP
Inherits \fBBlockCommand\fP\&.
.SS "Public Member Functions"

.in +1c
.ti -1c
.RI "\fBGetBroadcastMsg\fP (uint16_t abank, uint32_t auser, uint32_t block, uint32_t time)"
.br
.ti -1c
.RI "\fBGetBroadcastMsg\fP (const \fBGetBroadcastMsg\fP &obj)=delete"
.br
.RI "\fIDisabled copy contructor\&. \fP"
.ti -1c
.RI "\fBGetBroadcastMsg\fP & \fBoperator=\fP (const \fBGetBroadcastMsg\fP &)=delete"
.br
.RI "\fIDisabled copy assignment operator\&. \fP"
.ti -1c
.RI "virtual \fB~GetBroadcastMsg\fP ()"
.br
.RI "\fIFree responseBuffer\&. \fP"
.ti -1c
.RI "virtual int \fBgetType\fP () override"
.br
.RI "\fIReturn TXSTYPE_BLG as type \&. \fP"
.ti -1c
.RI "virtual CommandType \fBgetCommandType\fP () override"
.br
.RI "\fIReturn eReadingOnly as command type \&. \fP"
.ti -1c
.RI "virtual unsigned char * \fBgetData\fP () override"
.br
.RI "\fIGet pointer to command data structure\&. \fP"
.ti -1c
.RI "virtual unsigned char * \fBgetResponse\fP () override"
.br
.RI "\fIGet pointer to response data\&. \fP"
.ti -1c
.RI "virtual uint32_t \fBgetUserMessageId\fP () override"
.br
.RI "\fIGet message id\&. \fP"
.ti -1c
.RI "virtual void \fBsetData\fP (char *data) override"
.br
.RI "\fIPut data as a char list and put convert it to data structure\&. \fP"
.ti -1c
.RI "virtual void \fBsetResponse\fP (char *response) override"
.br
.RI "\fIApply data as a response struct\&. \fP"
.ti -1c
.RI "virtual int \fBgetDataSize\fP () override"
.br
.RI "\fIGet data struct size\&. Without signature\&. \fP"
.ti -1c
.RI "virtual int \fBgetResponseSize\fP () override"
.br
.RI "\fIGet response data struct size\&. \fP"
.ti -1c
.RI "virtual unsigned char * \fBgetSignature\fP () override"
.br
.RI "\fIGet pointer to signature data\&. \fP"
.ti -1c
.RI "virtual int \fBgetSignatureSize\fP () override"
.br
.RI "\fIGet signature size\&. \fP"
.ti -1c
.RI "virtual void \fBsign\fP (const uint8_t *hash, const uint8_t *sk, const uint8_t *pk) override"
.br
.RI "\fISign actual data plus hash using user private and public keys\&. \fP"
.ti -1c
.RI "virtual bool \fBcheckSignature\fP (const uint8_t *hash, const uint8_t *pk) override"
.br
.RI "\fICheck actual signature\&. \fP"
.ti -1c
.RI "virtual uint32_t \fBgetTime\fP () override"
.br
.RI "\fIGet time of command\&. \fP"
.ti -1c
.RI "virtual uint32_t \fBgetUserId\fP () override"
.br
.RI "\fIGet User ID\&. \fP"
.ti -1c
.RI "virtual uint32_t \fBgetBankId\fP () override"
.br
.RI "\fIGet Node ID\&. \fP"
.ti -1c
.RI "virtual int64_t \fBgetFee\fP () override"
.br
.RI "\fIGet command fee\&. \fP"
.ti -1c
.RI "virtual int64_t \fBgetDeduct\fP () override"
.br
.RI "\fIGet change in cash balance after command\&. \fP"
.ti -1c
.RI "virtual bool \fBsend\fP (\fBINetworkClient\fP &netClient) override"
.br
.RI "\fISend data to the server\&. \fP"
.ti -1c
.RI "virtual void \fBsaveResponse\fP (\fBsettings\fP &sts) override"
.br
.RI "\fISave command response to settings object\&. \fP"
.ti -1c
.RI "virtual std::string \fBtoString\fP (bool pretty) override"
.br
.ti -1c
.RI "virtual void \fBtoJson\fP (boost::property_tree::ptree &ptree) override"
.br
.ti -1c
.RI "virtual void \fBtxnToJson\fP (boost::property_tree::ptree &ptree) override"
.br
.ti -1c
.RI "virtual std::string \fBusageHelperToString\fP () override"
.br
.ti -1c
.RI "virtual void \fBreadDataBuffer\fP (unsigned char *dataBuffer, int size)"
.br
.ti -1c
.RI "virtual void \fBprintBlg\fP (\fBGetBroadcastResponse\fP &block, std::string &\fBmessage\fP, boost::property_tree::ptree &ptree)"
.br
.ti -1c
.RI "virtual bool \fBloadFromLocalPath\fP ()"
.br
.ti -1c
.RI "virtual void \fBsaveCopy\fP (unsigned char *dataBuffer, int size)"
.br
.ti -1c
.RI "uint32_t \fBgetBlockTime\fP ()"
.br
.in -1c
.SS "Public Attributes"

.in +1c
.ti -1c
.RI "\fBGetBroadcastData\fP \fBm_data\fP"
.br
.ti -1c
.RI "std::vector< std::pair< \fBGetBroadcastResponse\fP, std::string > > \fBm_response\fP"
.br
.ti -1c
.RI "\fBBroadcastFileHeader\fP \fBm_header\fP"
.br
.ti -1c
.RI "bool \fBm_loadedFromLocal\fP"
.br
.ti -1c
.RI "\fBcommandresponse\fP \fBm_fakeResponse\fP"
.br
.in -1c
.SH "Detailed Description"
.PP 
Get broadcast command class\&. 
.PP
Definition at line 11 of file getbroadcastmsg\&.h\&.
.SH "Member Function Documentation"
.PP 
.SS "bool GetBroadcastMsg::checkSignature (const uint8_t * hash, const uint8_t * pk)\fC [override]\fP, \fC [virtual]\fP"

.PP
Check actual signature\&. 
.PP
\fBParameters:\fP
.RS 4
\fIhash\fP Previous hash operation\&. 
.br
\fIpk\fP Pointer to public key\&. 
.RE
.PP

.PP
Implements \fBICommand\fP\&.
.PP
Definition at line 76 of file getbroadcastmsg\&.cpp\&.
.SS "bool GetBroadcastMsg::send (\fBINetworkClient\fP & netClient)\fC [override]\fP, \fC [virtual]\fP"

.PP
Send data to the server\&. 
.PP
\fBParameters:\fP
.RS 4
\fInetClient\fP Netwrok client implementation of \fBINetworkClient\fP interface\&. 
.br
\fIpk\fP Pointer to public key\&. 
.RE
.PP

.PP
Implements \fBICommand\fP\&.
.PP
Definition at line 103 of file getbroadcastmsg\&.cpp\&.
.SS "void GetBroadcastMsg::sign (const uint8_t * hash, const uint8_t * sk, const uint8_t * pk)\fC [override]\fP, \fC [virtual]\fP"

.PP
Sign actual data plus hash using user private and public keys\&. 
.PP
\fBParameters:\fP
.RS 4
\fIhash\fP Previous hash operation\&. 
.br
\fIsk\fP Pointer to provate key\&. 
.br
\fIpk\fP Pointer to public key\&. 
.RE
.PP

.PP
Implements \fBICommand\fP\&.
.PP
Definition at line 72 of file getbroadcastmsg\&.cpp\&.

.SH "Author"
.PP 
Generated automatically by Doxygen for esc from the source code\&.


CHANGELOG.md
# Changelog
All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/)
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## [Unreleased]

## [1.0.0] - 2018-08-24
### Changed
- Genesis file for the launch
- Update README

## [0.0.6] - 2018-08-10
### Added
- Profit sharing among TOP and VIP nodes

### Changed
- Genesis file for test launch

### Fixed
- Vote counting in get_block command
- Ability to send externally signed transactions

## [0.0.5] - 2018-08-01
### Added
- Support passing version by cmake command.
- Contributing guide
- "Dry run" test case

### Fixed
- Node stops after incorrect `send_many` request  - #134
- Fee checking
- Internal transactions numbers

## [0.0.4] - 2018-07-27
### Changed
- Update default parameters
- Creating a dev version with reference to the last tag

[Unreleased]: https://github.com/adshares/ads/compare/v1.0.0...HEAD

[1.0.0]: https://github.com/adshares/ads/compare/v0.0.6...v1.0.0
[0.0.6]: https://github.com/adshares/ads/compare/v0.0.5...v0.0.6
[0.0.5]: https://github.com/adshares/ads/compare/v0.0.4...v0.0.5
[0.0.4]: https://github.com/adshares/ads/compare/v0.0.3...v0.0.4


src/common/command/getbroadcastmsg.cpp
#include "getbroadcastmsg.h"

#include <iostream>
#include <fstream>
#include <sys/stat.h>

#include "ed25519/ed25519.h"
#include "abstraction/interfaces.h"
#include "helper/json.h"
#include "helper/hash.h"

GetBroadcastMsg::GetBroadcastMsg()
    : m_data{}, m_loadedFromLocal(false) {
    m_responseError = ErrorCodes::Code::eNone;
}

GetBroadcastMsg::GetBroadcastMsg(uint16_t abank, uint32_t auser, uint32_t block, uint32_t time)
    : m_data(abank, auser, block, time), m_loadedFromLocal(false) {
    m_responseError = ErrorCodes::Code::eNone;
}

GetBroadcastMsg::~GetBroadcastMsg() {
}

unsigned char* GetBroadcastMsg::getData() {
    return reinterpret_cast<unsigned char*>(&m_data.info);
}

unsigned char* GetBroadcastMsg::getResponse() {
    // only header
    return reinterpret_cast<unsigned char*>(&m_header);
}

void GetBroadcastMsg::setData(char* data) {
    m_data = *reinterpret_cast<decltype(m_data)*>(data);
}

void GetBroadcastMsg::setResponse(char* response) {
    // only header
    m_header = *reinterpret_cast<decltype(m_header)*>(response);
}

int GetBroadcastMsg::getDataSize() {
    return sizeof(m_data.info);
}

int GetBroadcastMsg::getResponseSize() {
    // only header
    return sizeof(m_header);
}

unsigned char* GetBroadcastMsg::getSignature() {
    return m_data.sign;
}

int GetBroadcastMsg::getSignatureSize() {
    return sizeof(m_data.sign);
}

int GetBroadcastMsg::getType() {
    return TXSTYPE_BLG;
}

CommandType GetBroadcastMsg::getCommandType() {
    return CommandType::eReadingOnly;
}

uint32_t GetBroadcastMsg::getUserMessageId() {
    return 0;
}

void GetBroadcastMsg::sign(const uint8_t* /*hash*/, const uint8_t* sk, const uint8_t* pk) {
    ed25519_sign(getData(), getDataSize(), sk, pk, getSignature());
}

bool GetBroadcastMsg::checkSignature(const uint8_t* /*hash*/, const uint8_t* pk) {
    return (ed25519_sign_open(getData(), getDataSize(), pk, getSignature()) == 0);
}

void GetBroadcastMsg::saveResponse(settings& /*sts*/) {
}

uint32_t GetBroadcastMsg::getUserId() {
    return m_data.info.src_user;
}

uint32_t GetBroadcastMsg::getBankId() {
    return m_data.info.src_node;
}

uint32_t GetBroadcastMsg::getTime() {
    return m_data.info.ttime;
}

int64_t GetBroadcastMsg::getFee() {
    return 0;
}

int64_t GetBroadcastMsg::getDeduct() {
    return 0;
}

bool GetBroadcastMsg::send(INetworkClient& netClient) {
    if (loadFromLocalPath()) {
        return true;
    }

    if(!netClient.sendData(getData(), sizeof(m_data))) {
        ELOG("GetBroadcastMsg sending error\n");
        return false;
    }

    if (!netClient.readData((int32_t*)&m_responseError, ERROR_CODE_LENGTH)) {
        ELOG("GetBroadcastMsg reading error\n");
        return false;
    }

    if (m_responseError) {
        return true;
    }

    if(!netClient.readData(getResponse(), getResponseSize())) {
        ELOG("GetBroadcastMsg reading header error\n");
        return false;
    }

    ELOG("size %ud %08X\n", m_header.fileSize, m_data.info.block);

    unsigned char *readBuffer = new unsigned char[m_header.fileSize];
    if (!netClient.readData(readBuffer, m_header.fileSize)) {
        ELOG("GetBroadcastMsg reading error\n");
        delete[] readBuffer;
        return false;
    }

    readDataBuffer(readBuffer, m_header.fileSize);

    saveCopy(readBuffer, m_header.fileSize);

    delete[] readBuffer;
    return true;
}

uint32_t GetBroadcastMsg::getBlockTime() {
    return m_data.info.block;
}

std::string GetBroadcastMsg::toString(bool /*pretty*/) {
    return "";
}

void GetBroadcastMsg::toJson(boost::property_tree::ptree& ptree) {
    if (m_responseError) {
        ptree.put(ERROR_TAG, ErrorCodes().getErrorMsg(m_responseError));
    } else {
        char blockhex[9];
        blockhex[8]='\0';
        sprintf(blockhex,"%08X", m_header.path);
        ptree.put("block_time_hex", blockhex);
        ptree.put("block_time", m_header.path);
        ptree.put("broadcast_count", m_response.size());
        if(m_response.size() > 0) {
            ptree.put("log_file", (m_loadedFromLocal) ? "archive" : "new");
            boost::property_tree::ptree blockTree;
            for (auto &it : m_response) {
                printBlg(it.first, it.second, blockTree);
            }

            ptree.add_child("broadcast", blockTree);
        }
    }
}

void GetBroadcastMsg::printBlg(GetBroadcastResponse& block, std::string& message, boost::property_tree::ptree &ptree) {
    char pkey[65];
    char hash[65];
    ed25519_key2text(pkey,block.data.pkey,32);
    ed25519_key2text(hash,block.data.hash,32);
    pkey[64]='\0';
    hash[64]='\0';
    uint16_t suffix=Helper::crc_acnt(block.info.src_node, block.info.src_user);
    char acnt[19];
    sprintf(acnt,"%04X-%08X-%04X", block.info.src_node, block.info.src_user, suffix);

    boost::property_tree::ptree blogentry;
    blogentry.put("block_time", m_header.path);
    blogentry.put("block_date", mydate(m_header.path));
    if (block.info.ttype != TXSTYPE_BRO) {
        ptree.put(ERROR_TAG, ErrorCodes().getErrorMsg(ErrorCodes::Code::eIncorrectTransaction));
        return;
    }
    blogentry.put("node", block.info.src_node);
    blogentry.put("account",block.info.src_user);
    blogentry.put("address",acnt);
    blogentry.put("account_msid", block.info.msg_id);
    blogentry.put("time", block.info.ttime);
    blogentry.put("date",mydate(block.info.ttime));

    //transaction data
    std::stringstream tx_data;
    Helper::ed25519_key2text(tx_data, reinterpret_cast<unsigned char*>(&block.info), sizeof(block.info));
    blogentry.put("data", tx_data.str());

    //message
    tx_data.str("");
    Helper::ed25519_key2text(tx_data, (unsigned char*)message.c_str(), block.info.msg_length);
    blogentry.put("message", tx_data.str());

    //signature
    tx_data.str("");
    Helper::ed25519_key2text(tx_data, reinterpret_cast<unsigned char*>(&block.data.sign), sizeof(block.data.sign));
    blogentry.put("signature", tx_data.str());

    //hash
    tx_data.str("");
    Helper::ed25519_key2text(tx_data, reinterpret_cast<unsigned char*>(&block.data.hash), sizeof(block.data.hash));
    blogentry.put("input_hash", tx_data.str());

    //public key
    tx_data.str("");
    Helper::ed25519_key2text(tx_data, reinterpret_cast<unsigned char*>(&block.data.pkey), sizeof(block.data.pkey));
    blogentry.put("public_key", tx_data.str());

    bool result = this->checkSignature(block.data.sign, block.data.pkey);
    blogentry.put("verify", (result) ? "passed" : "failed");

    //tx_id
    char tx_id[64];
    sprintf(tx_id,"%04X:%08X:%04X", block.info.src_node, block.data.msid, block.data.mpos);
    blogentry.put("node_msid", block.data.msid);
    blogentry.put("node_mpos", block.data.mpos);
    blogentry.put("id",tx_id);

    blogentry.put("fee",print_amount(TXS_BRO_FEE(block.info.msg_length)));
    ptree.push_back(std::make_pair("", blogentry));
}

bool GetBroadcastMsg::loadFromLocalPath() {
    uint32_t blockTime = this->getBlockTime();
    if (!blockTime) {
        return false;
    }

    uint32_t path = blockTime-(blockTime%BLOCKSEC);
    char filename[64];
    sprintf(filename, "bro/%08X.bin", path);
    std::ifstream file(filename, std::ifstream::binary | std::ifstream::in);
    if (!file.is_open()) {
        m_loadedFromLocal = false;
        return false;
    }

    struct stat sstat;
    stat(filename, &sstat);
    int remainData = sstat.st_size;
    m_header.fileSize = remainData;
    m_header.path = path;
    m_loadedFromLocal = true;

    while (remainData > 0) {
        GetBroadcastResponse response{};
        file.read((char*)&response.info, sizeof(response.info));
        remainData -= sizeof(response.info);

        char message[response.info.msg_length];
        file.read((char*)message, response.info.msg_length);
        remainData -= response.info.msg_length;

        file.read((char*)&response.data, sizeof(response.data));
        remainData -= sizeof(response.data);

        m_response.push_back(std::make_pair(response, std::string(message, response.info.msg_length)));
    }
    file.close();

    return true;
}

void GetBroadcastMsg::saveCopy(unsigned char *dataBuffer, int size) {
    uint32_t blockTime = m_header.path;
    if (!blockTime || m_response.empty() || size <= 0) {
        return;
    }

    uint32_t path = blockTime - blockTime%BLOCKSEC;

    char filename[64];
    mkdir("bro",0755);
    sprintf(filename, "bro/%08X.bin", path);
    std::ofstream file(filename, std::ofstream::binary | std::ofstream::out);
    if (file.is_open()) {
        file.write(reinterpret_cast<char*>(dataBuffer), size);
        file.close();
    }
}

void GetBroadcastMsg::readDataBuffer(unsigned char* dataBuffer, int size) {
    unsigned char *bufferPtr = dataBuffer;

    while (size > 0) {
        GetBroadcastResponse response{};

        // read SendBroadcastInfo data
        memcpy(&response.info, bufferPtr, sizeof(response.info));
        bufferPtr += sizeof(response.info);
        size -= sizeof(response.info);

        // read Message, depends on SendBroadcastInfo msg_length field
        char message[response.info.msg_length];
        memcpy(message, bufferPtr, response.info.msg_length);
        bufferPtr += response.info.msg_length;
        size -= response.info.msg_length;

        // read rest of file
        memcpy(&response.data, bufferPtr, sizeof(response.data));
        bufferPtr += sizeof(response.data);
        size -= sizeof(response.data);

        m_response.push_back(std::make_pair(response, std::string(message, response.info.msg_length)));
   }
}

void GetBroadcastMsg::txnToJson(boost::property_tree::ptree& ptree) {
    using namespace Helper;
    ptree.put(TAG::TYPE, getTxnName(m_data.info.ttype));
    ptree.put(TAG::SRC_NODE, m_data.info.src_node);
    ptree.put(TAG::SRC_USER, m_data.info.src_user);
    ptree.put(TAG::BLOCK, m_data.info.block);
    ptree.put(TAG::TIME, m_data.info.ttime);
    ptree.put(TAG::SIGN, ed25519_key2text(getSignature(), getSignatureSize()));
}

std::string GetBroadcastMsg::usageHelperToString() {
    std::stringstream ss{};
    ss << "Usage: " << "{\"run\":\"get_broadcast\",[\"from\":<timestamp>]}" << "\n";
    ss << "Example: " << "{\"run\":\"get_broadcast\",\"from\":\"1491210824\"}" << "\n";
    return ss.str();
}
